src/librbd/Watcher.cc

```diff
@@ -0,0 +1,233 @@
+    m_watch_handle = 0;
+  } else if (r >= 0) {
+    m_watch_state = WATCH_STATE_REGISTERED;
+  }
+}
+
+void Watcher::unregister_watch(Context *on_finish) {
+  ldout(m_cct, 10) << this << " unregistering watcher" << dendl;
+
+  RWLock::WLocker watch_locker(m_watch_lock);
+  if (m_watch_state == WATCH_STATE_REWATCHING) {
+    ldout(m_cct, 10) << this << " delaying unregister until rewatch completed"
+                     << dendl;
+
+    assert(m_unregister_watch_ctx == nullptr);
+    m_unregister_watch_ctx = new FunctionContext([this, on_finish](int r) {
+        unregister_watch(on_finish);
+      });
+    return;
+  }
+
+  if (m_watch_state == WATCH_STATE_REGISTERED ||
+      m_watch_state == WATCH_STATE_ERROR) {
+    m_watch_state = WATCH_STATE_UNREGISTERED;
+
+    librados::AioCompletion *aio_comp = create_rados_safe_callback(
+                      new C_UnwatchAndFlush(m_ioctx, on_finish));
+    int r = m_ioctx.aio_unwatch(m_watch_handle, aio_comp);
+    assert(r == 0);
+    aio_comp->release();
+  }
+}
+
+void Watcher::flush(Context *on_finish) {
+  m_notifier.flush(on_finish);
+}
+
+void Watcher::handle_error(uint64_t handle, int err) {
+  lderr(m_cct) << this << " watch failed: " << handle << ", "
+               << cpp_strerror(err) << dendl;
+
+  RWLock::WLocker l(m_watch_lock);
+  if (m_watch_state == WATCH_STATE_REGISTERED) {
+    m_watch_state = WATCH_STATE_ERROR;
+
+    FunctionContext *ctx = new FunctionContext(
+        boost::bind(&Watcher::rewatch, this));
+    m_work_queue->queue(ctx);
+  }
+}
+
+void Watcher::acknowledge_notify(uint64_t notify_id, uint64_t handle,
+	                                  bufferlist &out) {
+  m_ioctx.notify_ack(m_oid, notify_id, handle, out);
+}
+
+void Watcher::rewatch() {
+  ldout(m_cct, 10) << this << " re-registering watch" << dendl;
+
+  RWLock::WLocker l(m_watch_lock);
+  if (m_watch_state != WATCH_STATE_ERROR) {
+    return;
+  }
+  m_watch_state = WATCH_STATE_REWATCHING;
+
+  Context *ctx = create_context_callback<Watcher,
+                                         &Watcher::handle_rewatch>(this);
+  RewatchRequest *req = RewatchRequest::create(m_ioctx, m_oid, m_watch_lock,
+                                               &m_watch_ctx,
+                                               &m_watch_handle, ctx);
+  req->send();
+}
+
+void Watcher::handle_rewatch(int r) {
+  ldout(m_cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
+
+  WatchState next_watch_state = WATCH_STATE_REGISTERED;
+  if (r < 0) {
+    // only EBLACKLISTED or ENOENT can be returned
+    assert(r == -EBLACKLISTED || r == -ENOENT);
+    next_watch_state = WATCH_STATE_UNREGISTERED;
+  }
+
+  Context *unregister_watch_ctx = nullptr;
+  {
+    RWLock::WLocker watch_locker(m_watch_lock);
+    assert(m_watch_state == WATCH_STATE_REWATCHING);
+    m_watch_state = next_watch_state;
+
+    std::swap(unregister_watch_ctx, m_unregister_watch_ctx);
+
+    finish_rewatch(r);
```

Comment:
I am nervous about calling this while holding the lock. For example, ImageWatcher will attempt to reacquire the lock -- which means it needs to call ```get_watch_handle```.